File: python_examples/task5/problem_5.py
"""
2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?
"""


def is_divided_by_prime(numbers_to_check: list, number: int) -> bool:
    for n in numbers_to_check:
        if number % n != 0:
            return False
    return True


def get_smallest_number_divided_by_all(n: int):
    number = n
    start = 1 if n < 6 else 6
    numbers_to_check = [i for i in range(start, n)]
    while True:
        if is_divided_by_prime(numbers_to_check, number):
            return number
        else:
            number += n


print(f'Smallest number divided by all values to n is: {get_smallest_number_divided_by_all(20)}')

File: java-examples/src/com/szymczak/task5/Problem5.java
package com.szymczak.task5;

import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/*
 2520 is the smallest number that can be divided by each of the numbers
 from 1 to 10 without any remainder.
 What is the smallest positive number that is evenly divisible by all of
 the numbers from 1 to 20?
*/
public class Problem5 {
    public static void main(String[] args) {
        System.out.printf("Smallest number divided by all values to n is: %d\n", getSmallestNumberDividedByAll(20));
    }

    private static long getSmallestNumberDividedByAll(int i) {
        long tempI = i;
        int startPosition = i < 6 ? 1 : 6;
        List<Integer> numbersToCheck = IntStream
                .rangeClosed(startPosition, i)
                .boxed()
                .collect(Collectors.toList());
        System.out.println(numbersToCheck);

        while (true) {
            if (isDividedByAll(numbersToCheck, tempI)) {
                return tempI;
            }
            tempI += i;
        }
    }

    private static boolean isDividedByAll(List<Integer> numbersToCheck, long tempI) {
        for (Integer integer : numbersToCheck) {
            if (tempI % integer != 0) {
                return false;
            }
        }
        return true;
    }
}
